src/main/java/io/confluent/idesidecar/restapi/connections/DirectConnectionState.java

```diff
@@ -48,5 +78,126 @@ public Optional<Credentials> getSchemaRegistryCredentials(String clusterId) {
     return Optional.ofNullable(credentials);
   }
 
-  // TODO: DIRECT connections need validation checks
+  @Override
+  public Future<ConnectionStatus> getConnectionStatus() {
+    return Future.join(
+        getKafkaConnectionStatus(),
+        getSchemaRegistryConnectionStatus()
+    ).map(cf -> {
+      var futures = cf.list();
+      var kafkaStatus = (KafkaClusterStatus) futures.get(0);
+      var srStatus = (SchemaRegistryStatus) futures.get(1);
+      return ConnectionStatusBuilder
+          .builder()
+          .kafkaCluster(kafkaStatus)
+          .schemaRegistry(srStatus)
+          .build();
+    });
+  }
+
+  protected Future<KafkaClusterStatus> getKafkaConnectionStatus() {
+    var kafkaConfig = spec.kafkaClusterConfig();
+    if (kafkaConfig == null) {
+      return Future.succeededFuture(
+          ConnectionStatusKafkaClusterStatusBuilder
+              .builder()
+              .state(ConnectedState.NONE)
+              .build()
+      );
+    }
+    // There is a Kafka configuration, so validate the connection by creating an AdminClient
+    // and describing the cluster.
+    try (var adminClient = createAdminClient(kafkaConfig)) {
+      var clusterDesc = adminClient.describeCluster();
+      var actualClusterId = clusterDesc.clusterId().get(5, TimeUnit.SECONDS);
+      // Set the cluster ID
+      getSpec().kafkaClusterConfig().withId(actualClusterId);
+      return Future.succeededFuture(
+          ConnectionStatusKafkaClusterStatusBuilder
+              .builder()
+              .state(ConnectedState.SUCCESS)
+              .build()
+      );
+    } catch (Exception e) {
+      // The connection failed, so successfully return the failed state
+      return Future.succeededFuture(
+          ConnectionStatusKafkaClusterStatusBuilder
+              .builder()
+              .state(ConnectedState.FAILED)
+              .errors(
+                  new AuthErrors().withSignIn(
+                      "Failed to connect to Kafka cluster: %s".formatted(e.getMessage())
+                  )
+              ).build()
+      );
+    }
+  }
+
+  protected Future<SchemaRegistryStatus> getSchemaRegistryConnectionStatus() {
+
+    var schemaRegistryConfig = spec.schemaRegistryConfig();
+    if (schemaRegistryConfig == null) {
+      return Future.succeededFuture(
+          ConnectionStatusSchemaRegistryStatusBuilder
+              .builder()
+              .state(ConnectedState.NONE)
+              .build()
+      );
+    }
+    // There is a Schema Registry configuration, so validate the connection by creating a
+    // SchemaRegistryClient and getting the mode.
+    try (var srClient = createSchemaRegistryClient(schemaRegistryConfig)) {
```

Comment:
Do you know if `createSchemaRegistryClient` (or the constructor of `CachedSchemaRegistryClient`) uses a timeout when creating the SR client? If not, we'd probably wait for a long time when running into network issues so we want to set a timeout.

Alternatively, we could add timeouts to the futures in lines 84 and 85, and let them fail if they exceed the timeout. If we go down that road, we'd have to build the `FAILED` state in the method `getConnectionStatus()`.

I'm leaning towards the latter option. What's your take on it?

Reply:
The [default timeout](https://docs.confluent.io/platform/current/clients/confluent-kafka-dotnet/_site/api/Confluent.SchemaRegistry.CachedSchemaRegistryClient.html#Confluent_SchemaRegistry_CachedSchemaRegistryClient_DefaultTimeout) for all REST API calls made by the `CachedSchemaRegistryClient` is 30 seconds. It's configurable, so I can just reuse the same timeout value constant [discussed above](https://github.com/confluentinc/ide-sidecar/pull/169#discussion_r1846123371).